File: tests/fixtures/go/clean.go
// Clean fixture — real near-misses for every rule go.js implements, all of
// which must stay silent.
package clean

import (
	"crypto/rand"
	"crypto/sha256"
	"database/sql"
	"fmt"
	"log"
	"net/http"
	"os/exec"
)

func doWork() (int, error) {
	return 1, nil
}

func handlesError() error {
	result, err := doWork()
	if err != nil {
		return err
	}
	log.Println("worked", result)
	return nil
}

// The blank identifier here discards the loop value, not an error — the
// single highest false-positive risk for true/ignored-error.
func sumIndexes(xs []int) int {
	total := 0
	for i, _ := range xs {
		total += i
	}
	return total
}

func countKeys(m map[string]int) int {
	count := 0
	for k, _ := range m {
		count += len(k)
		_ = count
	}
	return count
}

func lookupUser(db *sql.DB, id string) (*sql.Rows, error) {
	return db.Query("SELECT * FROM t WHERE id = $1", id)
}

func runCommand(dir string) error {
	return exec.Command("ls", dir).Run()
}

func hashPassword(password string) [32]byte {
	return sha256.Sum256([]byte(password))
}

func makeToken() ([]byte, error) {
	buf := make([]byte, 16)
	_, err := rand.Read(buf)
	return buf, err
}

func mustNotPanic() error {
	return fmt.Errorf("something went wrong")
}

func fetchPage(url string) (*http.Response, error) {
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	return resp, nil
}

func startup() {
	log.Println("started")
}

File: tests/fixtures/rust/clean.rs
// Clean fixture — real near-misses for every rule rust.js implements, all of
// which must stay silent.

use std::process::Command;

fn parse(input: &str) -> Result<i32, std::num::ParseIntError> {
    input.parse::<i32>()
}

fn propagates(input: &str) -> Result<i32, std::num::ParseIntError> {
    let v = parse(input)?;
    Ok(v)
}

fn also_propagates(input: &str) -> Result<i32, std::num::ParseIntError> {
    parse(input)
}

// unwrap() on a Mutex lock is near-universal Rust practice: a poisoned lock
// means a prior panic already corrupted shared state, and there is no
// sensible recovery short of aborting — so procoder does not flag it.
fn read_counter(counter: &std::sync::Mutex<i32>) -> i32 {
    *counter.lock().unwrap()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_ok() {
        assert_eq!(parse("1").unwrap(), 1);
    }
}

// This function comes back after the test module in source order — it must
// still be treated as library code, not swept into the test region.
fn after_tests(input: &str) -> Result<i32, std::num::ParseIntError> {
    parse(input)
}

// SAFETY: p is non-null and points to a live, aligned i32, checked by the
// caller before this function is invoked.
unsafe fn read_raw(p: *const i32) -> i32 {
    unsafe { *p }
}

fn lookup_user(id: &str) {
    sqlx::query("SELECT * FROM t WHERE id = $1").bind(id);
}

fn run_it(dir: &str) {
    Command::new("ls").arg(dir).spawn().ok();
}

fn secure_client() {
    let _builder = reqwest::Client::builder();
}

fn make_token() -> u64 {
    use rand::rngs::OsRng;
    use rand::RngCore;
    OsRng.next_u64()
}

fn debug_print(x: i32) {
    tracing::info!("here {}", x);
}

/// Documentation that warns against a practice must not be flagged for the
/// practice: every rule rust.js has, named in prose, still silent.
///
/// - never `parse(input).unwrap()` or `.expect("should parse")` in a library
/// - never `sqlx::query(&format!("SELECT * FROM t WHERE id = {}", id))`
/// - never `Command::new("sh").arg("-c").arg(user_input)`
/// - never `.danger_accept_invalid_certs(true)`
/// - never `let token = rand::random::<u64>();` for a secret
/// - no leftover `println!("here")` or `dbg!(value)`
/// - never `unsafe { ptr::read(p) }` without a SAFETY note
pub fn documented_clean() {}
